### PR TITLE
Add app configuration health indicator

### PR DESCRIPTION
**Pull request by moarychan**
Fixes Azure/azure-sdk-for-java#25225

Changelog:
- Add app configuration health indicator
- Update actuator autoconfigure condition

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/cosmos/CosmosHealthConfiguration.java [hunk 1]
```diff
@@ -24,12 +24,12 @@
 @ConditionalOnClass({ CosmosAsyncClient.class, HealthIndicator.class})
 @ConditionalOnBean(CosmosAsyncClient.class)
 @AutoConfigureAfter(AzureCosmosAutoConfiguration.class)
-@ConditionalOnExpression("${spring.cloud.azure.cosmos.enabled:true}")
-@ConditionalOnProperty(prefix = "spring.cloud.azure.cosmos", name = { "endpoint", "database" })
+@ConditionalOnEnabledHealthIndicator("azure-cosmos")
+@ConditionalOnProperty(value = "spring.cloud.azure.cosmos.enabled", havingValue = "true", matchIfMissing = true)
+@ConditionalOnAnyProperty(prefix = "spring.cloud.azure.cosmos", name = "endpoint")
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/cosmos/CosmosHealthConfiguration.java:29]: It should also conditional on the property database?

File: sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/keyvault/KeyVaultHealthConfiguration.java [hunk 2]
```diff
@@ -22,10 +24,12 @@
 @ConditionalOnClass({ SecretClient.class, HealthIndicator.class })
 @ConditionalOnBean(SecretAsyncClient.class)
 @AutoConfigureAfter(AzureKeyVaultSecretAutoConfiguration.class)
+@ConditionalOnEnabledHealthIndicator("azure-key-vault")
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/keyvault/KeyVaultHealthConfiguration.java:27]: I prefer to call it azure-keyvault.

File: sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/appconfiguration/AppConfigurationHealthConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * Configuration class of App Configuration Health
+ */
+@Configuration
+@ConditionalOnClass({ ConfigurationClient.class, HealthIndicator.class })
+@ConditionalOnBean(ConfigurationClient.class)
+@AutoConfigureAfter(AzureAppConfigurationAutoConfiguration.class)
+@ConditionalOnEnabledHealthIndicator("azure-app-configuration")
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/appconfiguration/AppConfigurationHealthConfiguration.java:26]: I prefer to call it azure-appconfiguration.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/appconfiguration/AzureAppConfigurationProperties.java [hunk 4]
```diff
@@ -19,6 +19,7 @@
     private String connectionString;
     private ConfigurationServiceVersion serviceVersion;
 
+    @Override
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/appconfiguration/AzureAppConfigurationProperties.java:22]: Why is this also override?

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/cosmos/CosmosHealthConfiguration.java (hunk 1 above)
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/cosmos/CosmosHealthConfiguration.java:29]: It depends on `@AutoConfigureAfter(AzureCosmosAutoConfiguration.class)`, so it should be the same condition.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/appconfiguration/AzureAppConfigurationProperties.java (hunk 4 above)
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/appconfiguration/AzureAppConfigurationProperties.java]: Added the interface `ConnectionStringAware`

File: sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/cosmos/CosmosHealthConfiguration.java (hunk 1 above)
Comment (saragluna) [reply sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/cosmos/CosmosHealthConfiguration.java:29]: No, it's not the same, it's possible the AzureCosmosAutoConfiguration will not be enabled.

Comment (saragluna) [reply sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/cosmos/CosmosHealthConfiguration.java:29]: If we only need a @ConditionalOnBean(CosmosAsyncClient.class), we can skip these conditions: 
```java
@ConditionalOnProperty(value = "spring.cloud.azure.cosmos.enabled", havingValue = "true", matchIfMissing = true)
@ConditionalOnAnyProperty(prefix = "spring.cloud.azure.cosmos", name = "endpoint")
```

File: sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/appconfiguration/AppConfigurationHealthConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * Configuration class of App Configuration Health
+ */
+@Configuration
+@ConditionalOnClass({ ConfigurationAsyncClient.class, HealthIndicator.class })
+@ConditionalOnBean(ConfigurationAsyncClient.class)
+@AutoConfigureAfter(AzureAppConfigurationAutoConfiguration.class)
+@ConditionalOnEnabledHealthIndicator("azure-appconfiguration")
+@ConditionalOnProperty(value = "spring.cloud.azure.appconfiguration.enabled", havingValue = "true", matchIfMissing = true)
+@ConditionalOnAnyProperty(prefix = "spring.cloud.azure.appconfiguration", name = {"endpoint", "connection-string"})
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/appconfiguration/AppConfigurationHealthConfiguration.java:28]: Same here, if we have these two conditions 
```java
@ConditionalOnBean(ConfigurationAsyncClient.class)
@AutoConfigureAfter(AzureAppConfigurationAutoConfiguration.class)
```
we can skip these properties
```java
@ConditionalOnProperty(value = "spring.cloud.azure.appconfiguration.enabled", havingValue = "true", matchIfMissing = true)
@ConditionalOnAnyProperty(prefix = "spring.cloud.azure.appconfiguration", name = {"endpoint", "connection-string"})
```

File: sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/eventhubs/EventHubHealthConfiguration.java [hunk 6]
```diff
@@ -23,12 +25,14 @@
 @Configuration
 @ConditionalOnClass({ EventHubClientBuilder.class, HealthIndicator.class })
 @AutoConfigureAfter(AzureEventHubAutoConfiguration.class)
+@ConditionalOnEnabledHealthIndicator("azure-eventhubs")
+@ConditionalOnProperty(value = "spring.cloud.azure.eventhubs.enabled", havingValue = "true", matchIfMissing = true)
+@ConditionalOnAnyProperty(prefix = "spring.cloud.azure.eventhubs", name = { "connection-string", "namespace" })
 public class EventHubHealthConfiguration {
 
     @Bean
     @ConditionalOnBean(EventHubClientBuilder.class)
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/eventhubs/EventHubHealthConfiguration.java:34]: Is this condition not needed now? @ConditionalOnBean(EventHubClientBuilder.class)

File: sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/keyvault/KeyVaultHealthConfiguration.java [hunk 7]
```diff
@@ -22,11 +24,13 @@
 @ConditionalOnClass({ SecretClient.class, HealthIndicator.class })
 @ConditionalOnBean(SecretAsyncClient.class)
 @AutoConfigureAfter(AzureKeyVaultSecretAutoConfiguration.class)
+@ConditionalOnEnabledHealthIndicator("azure-keyvault")
+@ConditionalOnProperty(value = "spring.cloud.azure.keyvault.secret.enabled", havingValue = "true", matchIfMissing = true)
+@ConditionalOnAnyProperty(prefix = "spring.cloud.azure.keyvault.secret", name = "endpoint")
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/keyvault/KeyVaultHealthConfiguration.java:29]: Same here

File: sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/storage/StorageBlobHealthConfiguration.java [hunk 8]
```diff
@@ -21,10 +23,12 @@
 @ConditionalOnClass({ BlobServiceAsyncClient.class, HealthIndicator.class })
 @ConditionalOnBean(BlobServiceAsyncClient.class)
 @AutoConfigureAfter(AzureStorageBlobAutoConfiguration.class)
+@ConditionalOnEnabledHealthIndicator("azure-storage")
+@ConditionalOnProperty(value = "spring.cloud.azure.storage.blob.enabled", havingValue = "true", matchIfMissing = true)
+@ConditionalOnAnyProperty(prefix = "spring.cloud.azure.storage.blob", name = { "account-name", "endpoint", "connection-string" })
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/storage/StorageBlobHealthConfiguration.java:28]: same here

File: sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/storage/StorageFileHealthConfiguration.java [hunk 9]
```diff
@@ -21,10 +23,12 @@
 @ConditionalOnClass({ ShareServiceAsyncClient.class, HealthIndicator.class })
 @AutoConfigureAfter(AzureStorageFileShareAutoConfiguration.class)
 @ConditionalOnBean(ShareServiceAsyncClient.class)
+@ConditionalOnEnabledHealthIndicator("azure-storage")
+@ConditionalOnProperty(value = "spring.cloud.azure.storage.fileshare.enabled", havingValue = "true", matchIfMissing = true)
+@ConditionalOnAnyProperty(prefix = "spring.cloud.azure.storage.fileshare", name = { "account-name", "endpoint", "connection-string" })
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/storage/StorageFileHealthConfiguration.java:28]: same here

File: sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/storage/StorageQueueHealthConfiguration.java [hunk 10]
```diff
@@ -21,10 +23,12 @@
 @ConditionalOnClass({ QueueServiceAsyncClient.class, HealthIndicator.class })
 @ConditionalOnBean(QueueServiceAsyncClient.class)
 @AutoConfigureAfter(AzureStorageQueueAutoConfiguration.class)
+@ConditionalOnEnabledHealthIndicator("azure-storage")
+@ConditionalOnProperty(value = "spring.cloud.azure.storage.queue.enabled", havingValue = "true", matchIfMissing = true)
+@ConditionalOnAnyProperty(prefix = "spring.cloud.azure.storage.queue", name = { "account-name", "endpoint", "connection-string" })
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-actuator-autoconfigure/src/main/java/com/azure/spring/cloud/actuate/autoconfigure/storage/StorageQueueHealthConfiguration.java:28]: same here

File: sdk/spring/spring-cloud-azure-actuator/src/main/java/com/azure/spring/cloud/actuate/util/Constants.java [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+// Licensed under the MIT License.
+
+package com.azure.spring.cloud.actuate.util;
+
+/**
+ * Util class for Constants
+ */
+public class Constants {
+
+    public static final int DEFAULT_TIMEOUT_SECONDS = 3;
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-actuator/src/main/java/com/azure/spring/cloud/actuate/util/Constants.java:11]: How about changing this to Duration directly?

**Review by saragluna — CHANGES_REQUESTED** [review]

**Review by saragluna — APPROVED** [review]
LGTM.

**Review by saragluna — APPROVED** [review]
LGTM.

**Pull request merged**